pcap-dpdk.c

```diff
@@ -0,0 +1,742 @@
+	uint64_t bpf_drop;
+	struct ether_addr eth_addr;
+	struct timeval prev_ts;
+	struct rte_eth_stats prev_stats;
+	struct timeval curr_ts;
+	struct rte_eth_stats curr_stats;
+	uint64_t pps;
+	uint64_t bps;
+	struct rte_mempool * pktmbuf_pool;
+	struct dpdk_ts_helper ts_helper;
+	char pci_addr[DPDK_PCI_ADDR_SIZE];
+	unsigned char pcap_tmp_buf[RTE_ETH_PCAP_SNAPLEN];
+	volatile sig_atomic_t break_loop;
+};
+
+static struct rte_eth_conf port_conf = {
+	.rxmode = {
+		.split_hdr_size = 0,
+	},
+	.txmode = {
+		.mq_mode = ETH_MQ_TX_NONE,
+	},
+};
+
+static int dpdk_init_timer(struct pcap_dpdk *pd){
+	gettimeofday(&(pd->ts_helper.start_time),NULL);
+	pd->ts_helper.start_cycles = rte_get_timer_cycles();
+	pd->ts_helper.hz = rte_get_timer_hz();
+	if (pd->ts_helper.hz == 0){
+		return -1;	
+	}
+	return 0;
+}
+static inline void calculate_timestamp(struct dpdk_ts_helper *helper,struct timeval *ts)
+{
+	uint64_t cycles;
+	// delta
+	struct timeval cur_time;
+	cycles = rte_get_timer_cycles() - helper->start_cycles;
+	cur_time.tv_sec = (time_t)(cycles/helper->hz);
+	cur_time.tv_usec = (suseconds_t)((cycles%helper->hz)*1e6/helper->hz);
+	timeradd(&(helper->start_time), &cur_time, ts);
+}
+
+static unsigned int dpdk_gather_data(unsigned char *data, struct rte_mbuf *mbuf)
+{
+	unsigned int total_len = 0;
+	while (mbuf && (total_len+mbuf->data_len) < RTE_ETH_PCAP_SNAPLEN ){
+		rte_memcpy(data+total_len, rte_pktmbuf_mtod(mbuf,void *),mbuf->data_len);
+		total_len+=mbuf->data_len;
+		mbuf=mbuf->next;
+	}
+	return total_len;
+}
+
+static void pcap_dpdk_breakloop(pcap_t *p)
+{
+	pcap_breakloop_common(p);
+	struct pcap_dpdk *pd = (struct pcap_dpdk*)(p->priv);
+	pd->break_loop = p->break_loop;	
+}
+static void dpdk_dispatch_inter(void *dpdk_user)
+{
+	if (dpdk_user == NULL){
+		return;
+	}
+	pcap_t *p = dpdk_user;
+	struct pcap_dpdk *pd = (struct pcap_dpdk*)(p->priv);
+	int max_cnt = pd->max_cnt;
+	pcap_handler cb = pd->cb;
+	u_char *cb_arg = pd->cb_arg;
+	int nb_rx=0;
+	struct rte_mbuf *pkts_burst[MAX_PKT_BURST];
+	struct rte_mbuf *m;
+	struct pcap_pkthdr pcap_header;
+	uint16_t portid = pd->portid;
+	unsigned lcore_id = rte_lcore_id();
+	unsigned master_lcore_id = rte_get_master_lcore();
+	uint16_t data_len = 0;
+	u_char *bp = NULL;
+	int i=0;
+	unsigned int gather_len =0;
+	int pkt_cnt = 0;
+	int is_accepted=0;
+		
+	if(lcore_id == master_lcore_id){
+		RTE_LOG(INFO, USER1, "dpdk: lcoreid=%u runs for portid=%u\n", lcore_id, portid);
+	}else{
+		RTE_LOG(INFO, USER1, "dpdk: lcore %u has nothing to do\n", lcore_id);
+	}
+	//only use master lcore
+	if (lcore_id != master_lcore_id){
+		return;
+	}
+	while( max_cnt==-1 || pkt_cnt < max_cnt){
+		if (pd->break_loop){
+			break;
+		}
+		nb_rx = (int)rte_eth_rx_burst(portid, 0, pkts_burst, MAX_PKT_BURST);
+		pkt_cnt += nb_rx;
+		for ( i = 0; i < nb_rx; i++) {
+			m = pkts_burst[i];
+			calculate_timestamp(&(pd->ts_helper),&(pcap_header.ts));
+			data_len = rte_pktmbuf_data_len(m);
+			pcap_header.caplen = data_len; 
```

Comment:
You should probably cut the packet short by the snapshot length by setting `pcap_header.caplen` to the minimum of `data_len` and `p->snapshot`, so that if something's capturing packets with a snapshot length, and writing them to a file, they will write no more than the snapshot length to the file (that being one of the purposes of a snapshot length) without having to do anything other than writing out `pcap_header.caplen` bytes of packet data.

Reply:
Yes, caplen = min(pkt_len, p->snapshot).
I have modified the code.